File: olympic-hcm/2025-olympic11/hoctoan/hoctoan.py
import os
import sys

input_file = os.path.join(sys.path[0], 'hoctoan1.inp')
output_file = os.path.join(sys.path[0], 'hoctoan1.out')

a = b = 0
x = y = 0


def input_data():
    global a, b

    with open(input_file, 'r') as f:
        a, b = map(int, f.readline().split())
    

# Hàm tìm ước chung lớn nhất
def gcd(u, v):
    while u != 0:
        v, u = u, v % u

    return v


def process():
    global a, b, x, y
    
    if b % a != 0:
        x = -1
        return
    else:
        p = b / a

        y1 = 0

        # tổng x1 + y1 đang xét
        current_sum = 0

        # biến min_sum lưu tổng x1 + y1 nhỏ nhất
        min_sum = float('inf')

        # Duyệt từng ước x1 của p (p = b/a)
        for x1 in range(1, int(p ** 0.5) + 1):
            # Nếu x1 là ước của p
            if p % x1 == 0:
                # thì tìm ước còn lại y1
                y1 = p / x1

                # Nếu x1 và y1 nguyên tố cùng nhau
                if gcd(x1, y1) == 1:                
                    # thì tính tổng x1 + y1
                    current_sum = x1 + y1

                    # Cập nhật tổng nhỏ nhất
                    if current_sum < min_sum:
                        min_sum = current_sum
                        x = a * x1
                        y = a * y1
                    

def output():
    global x, y

    with open(output_file, 'w') as f:
        if x == -1:
            f.write(str(-1))        
        else:
            s = x + y
            f.write(f'{int(s)}\n{int(x)} {int(y)}')


if __name__ == '__main__':
    input_data()
    process()
    output()

File: olympic-hcm/2025-olympic11/hoctoan/hoctoan.cpp
#include <iostream>
#include <math.h>

#define input_file "hoctoan.inp"
#define output_file "hoctoan.out"

using namespace std;

int a, b;
int x, y;


void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    freopen(input_file, "r", stdin);
    
    cin >> a >> b;
}


// Hàm tìm ước chung lớn nhất
int gcd(int u, int v)
{
    while (u != 0)
    {
        int r = v % u;
        v = u;
        u = r;
    }

    return v;
}


void process()
{
    if (b % a != 0)
    {
        x = -1;
        return;
    }
    else
    {
        int p = b / a;

        int y1;

        // tổng x1 + y1 đang xét
        int current_sum;

        // biến min_sum lưu tổng x1 + y1 nhỏ nhất
        int min_sum = numeric_limits<int>::max();

        // Duyệt từng ước x1 của p (p = b/a)
        for (int x1 = 1; x1 * x1 < p; ++x1)
        {
            // Nếu x1 là ước của p
            if (p % x1 == 0)
            {
                // thì tìm ước còn lại y1
                y1 = p / x1;

                // Nếu x1 và y1 nguyên tố cùng nhau
                if (gcd(x1, y1) == 1)
                {
                    // thì tính tổng x1 + y1
                    current_sum = x1 + y1;

                    // Cập nhật tổng nhỏ nhất
                    if (current_sum < min_sum)
                    {
                        min_sum = current_sum;
                        x = a * x1;
                        y = a * y1;
                    }
                }
            }
        }
    }
}


void output()
{
    freopen(output_file, "w", stdout);

    if (x == -1)
    {
        cout << -1;
    }
    else
    {
        int s = x + y;
        cout << s << '\n'; 
        cout << x << ' ' << y;  
    }
}

int main()
{
    input();
    process();
    output();

    return 0;
}
